### PR TITLE
Add Hypergeometric Distribution to Distribution Library (WIP)

### PR DESCRIPTION
**Pull request by Harivallabha**
Add HyperGeometric distribution to the Discrete distributions module. Work in progress.

### CODE REVIEW

File: pymc3/distributions/discrete.py [hunk 1]
```diff
@@ -792,6 +792,27 @@ def _repr_latex_(self, name=None, dist=None):
         return r'${} \sim \text{{Geometric}}(\mathit{{p}}={})$'.format(name,
                                                 get_variable_name(p))
 
+class HyperGeometeric(Discrete):
+
```
Comment (ferrine) [inline pymc3/distributions/discrete.py:844]: Missing a docstring

Comment (Harivallabha) [reply pymc3/distributions/discrete.py:844]: Just added a docstring

File: pymc3/distributions/discrete.py [hunk 2]
```diff
@@ -793,6 +793,53 @@ def _repr_latex_(self, name=None, dist=None):
                                                 get_variable_name(p))
 
 class HyperGeometeric(Discrete):
+	 R"""
```
Comment (twiecki) [inline pymc3/distributions/discrete.py]: Bad indent.

File: pymc3/distributions/discrete.py [hunk 3]
```diff
@@ -823,21 +899,22 @@ class DiscreteUniform(Discrete):
...
-        for l, u in zip(ls, us):
-            x = np.arange(l, u+1)
-            pmf = [1.0 / (u - l + 1)] * len(x)
-            plt.plot(x, pmf, '-o', label='lower = {}, upper = {}'.format(l, u))
-        plt.xlabel('x', fontsize=12)
-        plt.ylabel('f(x)', fontsize=12)
-        plt.ylim(0, 0.4)
-        plt.legend(loc=1)
-        plt.show()
+		import matplotlib.pyplot as plt
```
Comment (twiecki) [inline pymc3/distributions/discrete.py]: These and the below shouldn't show up, or were these broken?

Comment (Harivallabha) [reply pymc3/distributions/discrete.py]: Oh damn sorry, yeah you're right. Messed that up, will change it.

File: pymc3/distributions/discrete.py (hunk 2 above)
Comment (Harivallabha) [reply pymc3/distributions/discrete.py]: Corrected that in the subsequent commit.

File: pymc3/distributions/discrete.py [hunk 4]
```diff
@@ -792,6 +792,115 @@ def _repr_latex_(self, name=None, dist=None):
...
+            values are desired the values must be provided in a numpy array or theano tensor
+
+        Returns
+        -------
+        TensorVariable
+        """
+        k = self.k
+        N = self.N
+        n = self.n
+        return bound(binomln(k, value) + binomln(N-k, value) - binomln(N,n),
```
Comment (twiecki) [inline pymc3/distributions/discrete.py]: pep8 requires white-spaces around operators and after commas.

File: pymc3/distributions/discrete.py [hunk 5]
```diff
@@ -792,6 +792,115 @@ def _repr_latex_(self, name=None, dist=None):
...
+                     0 <= k, k <= N, 0<=n, 0<=N, max(0, n - N + k) <=value, value<=min(k,n))
```
Comment (twiecki) [inline pymc3/distributions/discrete.py]: same.

File: pymc3/distributions/discrete.py [hunk 6]
```diff
@@ -845,6 +954,9 @@ def _random(self, lower, upper, size=None):
         samples = stats.randint.rvs(lower, upper + 1, size=size)
         return samples
 
+
```
Comment (twiecki) [inline pymc3/distributions/discrete.py]: can remove these.

**Comment by twiecki** [comment]
Also needs a line in release-notes.

File: pymc3/distributions/discrete.py [hunk 7]
```diff
@@ -792,6 +792,115 @@ def _repr_latex_(self, name=None, dist=None):
...
+        Total size of the population
+    n : integer
+    	Number of samples drawn from the population
+    k : integer
+    	Number of successful individuals in the population
+    """
+
+    def __init__(self, N,  k, n, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.k = k = tt.as_tensor_variable(floatX(k))
```
Comment (lucianopaz) [inline pymc3/distributions/discrete.py]: These need to be `intX` instead of `floatX`

File: pymc3/distributions/discrete.py [hunk 8]
```diff
@@ -792,6 +792,115 @@ def _repr_latex_(self, name=None, dist=None):
...
+            conditioned (uses default point if not specified).
+        size : int, optional
+            Desired size of random sample (returns one sample if not
+            specified).
+
+        Returns
+        -------
+        array
+        """
+        N,n,k = draw_values([self.N, self.n, self.k], point=point, size=size)
```
Comment (lucianopaz) [inline pymc3/distributions/discrete.py]: Spaces after commas, as in pep8

**Review by lucianopaz — CHANGES_REQUESTED** [review]
I agree with @twiecki. But this also needs a test of the `random` method in test_distributions_random.py and a test of the `logp` in test_distributions.py.

File: pymc3/distributions/discrete.py (hunk 4 above)
Comment (Harivallabha) [reply pymc3/distributions/discrete.py]: Done, thanks!

File: pymc3/distributions/discrete.py (hunk 5 above)
Comment (Harivallabha) [reply pymc3/distributions/discrete.py]: Done, thanks!

File: pymc3/distributions/discrete.py (hunk 7 above)
Comment (Harivallabha) [reply pymc3/distributions/discrete.py]: Done, thanks!

File: pymc3/distributions/discrete.py [hunk 9]
```diff
@@ -791,6 +791,116 @@ def _repr_latex_(self, name=None, dist=None):
...
+    k : integer
+        Number of successful individuals in the population
+    """
+
+    def __init__(self, N,  k, n, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.k = k = tt.as_tensor_variable(intX(k))
+        self.N = N = tt.as_tensor_variable(intX(N))
+        self.n = n = tt.as_tensor_variable(intX(n))
+        self.mode = 1
```
Comment (AustinRochford) [inline pymc3/distributions/discrete.py:850]: Implement the actual mode calculation

File: pymc3/distributions/discrete.py [hunk 10]
```diff
@@ -791,6 +791,116 @@ def _repr_latex_(self, name=None, dist=None):
...
+            specified).
+
+        Returns
+        -------
+        array
+        """
+        N, n, k = draw_values([self.N, self.n, self.k], point=point, size=size)
+        return generate_samples(np.random.hypergeometric, N, n, k,
+                                dist_shape=self.shape,
+                                size=size)
```
Comment (AustinRochford) [inline pymc3/distributions/discrete.py:872]: Per PEP8 there should be an empty line here

**Review by AustinRochford — CHANGES_REQUESTED** [review]

**Comment by twiecki** [comment]
Closing due to inactivity, feel free to reopen.

File: pymc3/distributions/discrete.py [hunk 11]
```diff
@@ -791,6 +791,116 @@ def _repr_latex_(self, name=None, dist=None):
...
+            values are desired the values must be provided in a numpy array or theano tensor
+
+        Returns
+        -------
+        TensorVariable
+        """
+        k = self.k
+        N = self.N
+        n = self.n
+        return bound(binomln(k, value) + binomln(N - k, value) - binomln(N, n),
```
Comment (ricardoV94) [inline pymc3/distributions/discrete.py:890]: I think there is an error in this line: the second binomial term should be binomln(N - k, n - value)

**Comment by ricardoV94** [comment]
Any reason why this PR was dropped? Could we reopen-it?

**Comment by twiecki** [comment]
@ricardoV94 If you want to continue to shepherd it through that would be great. Best to take this branch, fork it, and reopen a new PR.

**Pull request closed without merging**